### ISSUE TITLE
SliceReadoutTime range checking

### ISSUE THREAD

**Issue by jan-petr** [ExploreASL/ExploreASL#397]
SliceReadoutTime in parms2BIDS is reported as out of range. This is obvious for a vector that start with 0, which is always out of range. Solution: For SliceReadoutTime and a vector - make a subtraction of the first two fields before checking the range.

### Release note
Doesn't need a special mention in release changes and the issue number can be added to the SliceReadoutTime revamping

**Comment by HenkMutsaerts** [comment]
Not really sure, this is a check over multiple parameters, so it would be better to have fewest exceptions as possible. See my proposition below to create a new parameter SliceReadoutTime (which is a vector with the diff of SliceTiming), that could be checked similar to the other parameters.

* All SliceTiming's should be > 1 right? All SliceTimings are converted to the new SliceTiming right? Or is it throughout the pipeline having two ways to use SliceTiming (that would be really non-transparent...)
* The SliceTiming potential range is 0 to e.g., 10 (as this is in seconds) or 10000 (if this is in ms).
* But we create a new parameter SliceReadoutTime (which is simply the difference of SliceTiming)
  Then you can keep how it was right?

**Comment by jan-petr** [comment]
> Not really sure, this is a check over multiple parameters, so it would be better to have fewest exceptions as possible. See my proposition below to create a new parameter SliceReadoutTime (which is a vector with the diff of SliceTiming), that could be checked similar to the other parameters.

This is good, but potentially problematic. To avoid errors, we want to directly check the converted structure. So SliceReadoutTime would have to be added to the converted x-struct for checking. But that would be a redundant information and potential for errors. We would have to later delete it. So I would keep the current check with exception, rather than calculating a special field just for the check. But see also other comments below.

> * All SliceTiming's should be > 1 right? All SliceTimings are converted to the new SliceTiming right? Or is it throughout the pipeline having two ways to use SliceTiming (that would be really non-transparent...)

SliceTiming/SliceReadoutTime can be both inputted as a scalar or a full vector. And only later in the quantification, it is transformed to a vector in all cases and directly applied. So some parts of the vector can be <= 1. So not really non-transparent - both possibilities of entry (scalar and vector are possible). This comes from the fact that in legacy, scalar was inputted, and BIDS requires a vector. So I made both possible and checked and converted before application using a single function. However, we can/should move this conversion already to parms2BIDS. But lets make that a new issue.

> * The SliceTiming potential range is 0 to e.g., 10 (as this is in seconds) or 10000 (if this is in ms).

Yes.

> * But we create a new parameter SliceReadoutTime (which is simply the difference of SliceTiming)

Yes. This would work, but we would have an extra redundant parameter. Note that SliceTiming is always a vector (per BIDS). SliceReadoutTime can be both scalar (legacy) and vector (legacy to accommodate multi-band). So we can't have one always a vector and the other always a scalar.

>   Then you can keep how it was right?

My proposition - keep code as it was. But indeed make sure that the SliceTiming is always a vector, which is now not the case when SliceReadoutTime is converted to SliceTiming, because only ms->s conversion is applied and vector conversion is done only before quantification.

**Comment by HenkMutsaerts** [comment]
Apologies for verbose answer, otherwise let's discuss if needed :)

>> This is good, but potentially problematic. To avoid errors, we want to directly check the converted structure. So SliceReadoutTime would have to be added to the converted x-struct for checking. But that would be a redundant information and potential for errors. We would have to later delete it. So I would keep the current check with exception, rather than calculating a special field just for the check. But see also other comments below.

With all this code changing etc, I'm a big fan of "self-explanatory" names (as you know ;) so I would really not use a BIDS definition with slightly different behavior internally, then I would just use SliceReadoutTime internally, as we had it, and add extra functionality/checks, and proper conversion of `SliceTime->SliceReadoutTime` always, until we remove SliceReadoutTime internally and fully replace it by `SliceTime`. Like we now always have ms internally and seconds in BIDS, until we fully replace ms internally with s.


>> SliceTiming/SliceReadoutTime can be both inputted as a scalar or a full vector. And only later in the quantification, it is transformed to a vector in all cases and directly applied. So some parts of the vector can be <= 1. So not really non-transparent - both possibilities of entry (scalar and vector are possible). This comes from the fact that in legacy, scalar was inputted, and BIDS requires a vector. So I made both possible and checked and converted before application using a single function. However, we can/should move this conversion already to parms2BIDS. But let's make that a new issue.

"BIDSifying ExploreASL internally" is something we need to do in general, I'm fine with waiting with this. I just wanted to be sure that everything is "transparent and modular", so the conversion of both to the vector SliceReadoutTime that we use internally should be done at one location. Is that the case now?

>> This would work, but we would have an extra redundant parameter. Note that SliceTiming is always a vector (per BIDS). SliceReadoutTime can be both scalar (legacy) and vector (legacy to accommodate multi-band). So we can't have one always a vector and the other always a scalar.

True. I would just ensure that throughout the pipeline, SliceReadoutTime is always the difference of SliceTiming, just that we always use the same definitions. Then we can indeed later convergence to BIDS. Is it internally somewhere checked that SliceReadoutTime must be a single value if no multi-Band is requested? is "multi-band" recognized by a separate parameter? Or is "multi-band" assumed by length(SliceReadoutTime) > 1? I'm fine with the latter as well, but this should be clear (comments in code & verbose warning), and we need to make sure that this doesn't potentially hurt the standard user's quantification. Remember that it's more important to cater the 95% of non-multi-band, so in my view a length(SliceReadoutTime)>1 should at least always give a warning, e.g., `"WARNING: Multiple SliceReadoutTime values found, assuming multi-band readout!"`

>> My proposition - keep code as it was. But indeed make sure that the SliceTiming is always a vector, which is now not the case when SliceReadoutTime is converted to SliceTiming, because only ms->s conversion is applied and vector conversion is done only before quantification.

This confuses me: conversion should always be `SliceTiming (BIDS) -> SliceReadoutTime (ExploreASL internally)` right? We don't use SliceTiming internally (because if we do, it should always be fully BIDS compatible). I would either use a definition that is fully BIDS compatible, or give it a different name (like SliceReadoutTime) to make clear that this has a (potentially) different behavior than the BIDS definition.

Would it be clearer to do the vector conversion at the same location as the ms->s conversion (`xASL_adm_LoadParms`)?
Could still be a different function, and I'm fine calling this in e.g., `xASL_quant_*` as well, but it should be clear internally when parms are loaded?

**Comment by jan-petr** [comment]
> Apologies for verbose answer, otherwise let's discuss if needed :)

Haha, no worries - the problem is that this is really a bit more complicated. So that's why there's the name confusion etc. I'll explain below and then we can decide on the correct naming. We have now several behaviors in the transition state.
1. ASL.json - we use BIDS fields, SliceTiming always as a vector in seconds.
2. DataPar.json - this is still in legacy, so we have SliceReadoutTime in ms and scalar, but vector is now also allowed.
3. x-struct - internally, we use legacy - SliceReadoutTime in ms. Both vector and scalar is allowed, as no distinction is currently made between reading it in BIDS or Legacy - we could split those and save SliceReadoutTime in scalar and ms, and SliceTiming in s and vector, but that would be perhaps way too confusing. And the ms/s is more important than vector/scalar IMHO. The important thing is that ASL.json have always the full BIDS format with a vector.
4. Inside the wrappers, just before the application we call a single function that takes the x.SliceReadoutTime and prepares it for application. That means keeps it in ms and makes a vector out of it. Note that we should use this conversion always before application instead of saving back to x-struct because x-struct can hold only a single x.SliceReadoutTime, but this does differ between M0 and ASL in a few datasets. It is true that the converted name SliceTiming is slightly incorrect - as it is according to BIDS a vector, but it is in MS and not S. 

So point 1-3 is perfectly fine. Point 4 - we can either rename the variable to SliceReadoutTime or convert it to S and make sure that upon application S and not MS is used. Up to you - both are fine to me. However, since all occurrences in JSONs and x-struct are fine with BIDS, and the only problem is an unsaved variable, I don't think that this is really a huge problem.

> > > This is good, but potentially problematic. To avoid errors, we want to directly check the converted structure. So SliceReadoutTime would have to be added to the converted x-struct for checking. But that would be a redundant information and potential for errors. We would have to later delete it. So I would keep the current check with exception, rather than calculating a special field just for the check. But see also other comments below.
> 
> With all this code changing etc, I'm a big fan of "self-explanatory" names (as you know ;) so I would really not use a BIDS definition with slightly different behavior internally, then I would just use SliceReadoutTime internally, as we had it, and add extra functionality/checks, and proper conversion of `SliceTime->SliceReadoutTime` always, until we remove SliceReadoutTime internally and fully replace it by `SliceTime`. Like we now always have ms internally and seconds in BIDS, until we fully replace ms internally with s.

I agree - this confusion is caused by the fact that the variable is a vector (so SliceTiming is more fitting) and in ms (so SliceReadoutTime is more fitting). See above...

> > > SliceTiming/SliceReadoutTime can be both inputted as a scalar or a full vector. And only later in the quantification, it is transformed to a vector in all cases and directly applied. So some parts of the vector can be <= 1. So not really non-transparent - both possibilities of entry (scalar and vector are possible). This comes from the fact that in legacy, scalar was inputted, and BIDS requires a vector. So I made both possible and checked and converted before application using a single function. However, we can/should move this conversion already to parms2BIDS. But let's make that a new issue.
> 
> "BIDSifying ExploreASL internally" is something we need to do in general, I'm fine with waiting with this. I just wanted to be sure that everything is "transparent and modular", so the conversion of both to the vector SliceReadoutTime that we use internally should be done at one location. Is that the case now?

More complex. We don't BIDSify as such. We read it from JSONs and make sure that the units are converted (xASL_bids_parms2BIDS). But we can't fully BIDSify there - because for scalar->vector transform, we need to know the number of slices. We can't do that on a common place in xASL_module, as this might differ for ASL and M0. So the only solution is to make sure that stuff are BIDSified just before use. But there's a single function now doing it for all (does that fit you definition of single place? :) Single function, or a single location?)

> > > This would work, but we would have an extra redundant parameter. Note that SliceTiming is always a vector (per BIDS). SliceReadoutTime can be both scalar (legacy) and vector (legacy to accommodate multi-band). So we can't have one always a vector and the other always a scalar.

> True. I would just ensure that throughout the pipeline, SliceReadoutTime is always the difference of SliceTiming, just that we always use the same definitions. Then we can indeed later convergence to BIDS. Is it internally somewhere checked that SliceReadoutTime must be a single value if no multi-Band is requested? is "multi-band" recognized by a separate parameter? Or is "multi-band" assumed by length(SliceReadoutTime) > 1? I'm fine with the latter as well, but this should be clear (comments in code & verbose warning), and we need to make sure that this doesn't potentially hurt the standard user's quantification. Remember that it's more important to cater the 95% of non-multi-band, so in my view a length(SliceReadoutTime)>1 should at least always give a warning, e.g., `"WARNING: Multiple SliceReadoutTime values found, assuming multi-band readout!"`

That would be very tricky. You could have multi-band, or decreasing slice order etc - plenty of possibilities where this could be different. It would be too complex checking all the cases. I would stick to the basics - SliceReadoutTime MS and vector or scalar, SliceTime S and always vector. No need to restrict SliceReadoutTime to scalar - that would be unnecessarily restrictive. Besides - this behavior is accounted for, properly checked, correctly converted, so there's really no problem with it and no warnings necessary.

> > > My proposition - keep code as it was. But indeed make sure that the SliceTiming is always a vector, which is now not the case when SliceReadoutTime is converted to SliceTiming, because only ms->s conversion is applied and vector conversion is done only before quantification.
> 
> This confuses me: conversion should always be `SliceTiming (BIDS) -> SliceReadoutTime (ExploreASL internally)` right? We don't use SliceTiming internally (because if we do, it should always be fully BIDS compatible). I would either use a definition that is fully BIDS compatible, or give it a different name (like SliceReadoutTime) to make clear that this has a (potentially) different behavior than the BIDS definition.

Yes. We always convert SliceTiming (BIDS) to SliceReadoutTime (legacy) upon reading JSONs. But note that we have both legacy scalar inputs and BIDS vector inputs. Always internally converting everything to scalars would be unnecessary and prone to bugs. That's why we have vectors and scalar permitted for SliceReadoutTime.

> Would it be clearer to do the vector conversion at the same location as the ms->s conversion (`xASL_adm_LoadParms`)?
> Could still be a different function, and I'm fine calling this in e.g., `xASL_quant_*` as well, but it should be clear internally when parms are loaded?

Yes we do ms->s at xASL_bids_parms2BIDS, but then before calling xASL_quant functions, we have to make sure that it is always a vector. We had it inside xASL_quant, but you wanted to have a single function - so we have that now :)

To sum up. All is fine. All JSONs a x-structs are fine. We just have to decide if we want to keep the single variable as SliceTiming or SliceReadoutTime - but that's just a single internat variable...

**Comment by HenkMutsaerts** [comment]
You can only call it SliceTiming (==the BIDS name) internally, if it in all cases is exactly according to the BIDS definition. If not, then let's call it SliceReadoutTime internally, to make clear that it represents something different from the BIDS parameter.

And most important: to include comments and headers etc., such that it is clear for others to work with :)

**Comment by HenkMutsaerts** [comment]
>> Yes, only issue is the ASL-M0 difference. And x-struct unfortunately only holds a single general value and does not make this distinction.

We could rename both parameters by e.g., adding suffixes to `SliceReadoutTime_ASL` and `SliceReadoutTime_M0`, if we want to avoid this?

>> Not completely. In 95% cases where defined through DataPar.json, it is a scalar, and only Wibeke has it as a vector.
New import - both BIDS and non-BIDS now reads SliceReadoutTime from DICOM and saves it as a SliceTiming to BIDS in the correct format (vector and s). So for around 70% of datasets, this is now read and saved correctly into ASL4D.json. It gets it from many DICOMs, not from enhanced DICOMs. When importing from ASL4D.json to Legacy, I don't do this check because I would have to report nearly all data - warning makes no sense to me here. We only warn when differing from the number of slices.

Sure, but still in most cases where `SliceTiming` is a vector, `SliceReadoutTime` will be a scalar (unless we move fully to BIDS). When we move fully to BIDS, the warning should still be there, but then if `length(unique(diff(SliceTiming(:)))>1`.
In other words, if the timing difference between slices varies between slices, this should issue a warning, as this is only the case in 5% of cases.

Otherwise I'm happy :)
